controller/kubernetes_controller_test.go

```diff
@@ -0,0 +1,497 @@
+	// pv unset
+	tc = generateKubernetesTestCaseTemplate()
+	tc.copyCurrentToExpect()
+	tc.pv = nil
+	tc.pod.Status.Phase = apiv1.PodRunning
+	tc.expectVolume.Status.KubernetesStatus = types.KubernetesStatus{}
+	testCases["pv unset"] = tc
+
+	// pvc unset
+	tc = generateKubernetesTestCaseTemplate()
+	tc.copyCurrentToExpect()
+	tc.pv.Status.Phase = apiv1.VolumeAvailable
+	tc.pv.Spec.ClaimRef = nil
+	tc.pvc = nil
+	tc.expectVolume.Status.KubernetesStatus = types.KubernetesStatus{
+		PVName:       TestPVName,
+		PVStatus:     string(apiv1.VolumeAvailable),
+		LastPVCRefAt: "",
+		LastPodRefAt: "",
+		Namespace:    "",
+		PVCName:      "",
+		PodName:      "",
+		PodStatus:    "",
+		WorkloadName: "",
+		WorkloadType: "",
+	}
+	testCases["pvc unset"] = tc
+
+	// pod unset
+	tc = generateKubernetesTestCaseTemplate()
+	tc.copyCurrentToExpect()
+	tc.pv.Status.Phase = apiv1.VolumeBound
+	tc.pod = nil
+	tc.expectVolume.Status.KubernetesStatus = types.KubernetesStatus{
+		PVName:       TestPVName,
+		PVStatus:     string(apiv1.VolumeBound),
+		LastPVCRefAt: "",
+		LastPodRefAt: "",
+		Namespace:    TestNamespace,
+		PVCName:      TestPVCName,
+		PodName:      "",
+		PodStatus:    "",
+		WorkloadName: "",
+		WorkloadType: "",
+	}
+	testCases["pod unset"] = tc
+
+	// workload unset
+	tc = generateKubernetesTestCaseTemplate()
+	tc.copyCurrentToExpect()
+	tc.pv.Status.Phase = apiv1.VolumeBound
+	tc.pod.Status.Phase = apiv1.PodRunning
+	tc.pod.ObjectMeta.OwnerReferences = nil
+	tc.expectVolume.Status.KubernetesStatus = types.KubernetesStatus{
+		PVName:       TestPVName,
+		PVStatus:     string(apiv1.VolumeBound),
+		LastPVCRefAt: "",
+		LastPodRefAt: "",
+		Namespace:    TestNamespace,
+		PVCName:      TestPVCName,
+		PodName:      TestPodName,
+		PodStatus:    string(apiv1.PodRunning),
+		WorkloadName: "",
+		WorkloadType: "",
+	}
+	testCases["workload unset"] = tc
+
+	// pod phase updated: running -> failed
+	tc = generateKubernetesTestCaseTemplate()
+	tc.pv.Status.Phase = apiv1.VolumeBound
+	tc.pod.Status.Phase = apiv1.PodRunning
+	tc.podStatusUpdated = string(apiv1.PodFailed)
+	tc.copyCurrentToExpect()
+	tc.expectVolume.Status.KubernetesStatus = types.KubernetesStatus{
+		PVName:       TestPVName,
+		PVStatus:     string(apiv1.VolumeBound),
+		LastPVCRefAt: "",
+		LastPodRefAt: "",
+		Namespace:    TestNamespace,
+		PVCName:      TestPVCName,
+		PodName:      TestPodName,
+		PodStatus:    string(apiv1.PodFailed),
+		WorkloadName: TestWorkloadName,
+		WorkloadType: TestWorkloadKind,
+	}
+	testCases["pod phase updated to 'failed'"] = tc
+
+	// pod deleted
+	tc = generateKubernetesTestCaseTemplate()
+	tc.pv.Status.Phase = apiv1.VolumeBound
+	tc.pod.Status.Phase = apiv1.PodRunning
+	tc.podStatusUpdated = TestStatusDeleted
+	tc.copyCurrentToExpect()
+	tc.expectVolume.Status.KubernetesStatus = types.KubernetesStatus{
+		PVName:       TestPVName,
+		PVStatus:     string(apiv1.VolumeBound),
+		LastPVCRefAt: "",
+		// set <LastPodRefAt> later since it is a timestamp
```

Comment:
See https://github.com/rancher/longhorn-manager/blob/master/controller/volume_controller_test.go#L74 and https://github.com/rancher/longhorn-manager/blob/master/controller/volume_controller_test.go#L208 on how timestamp are handled.